internal/exec/stages/files/units.go

```diff
@@ -19,35 +19,66 @@ import (
 	"path/filepath"
 	"strings"
 
+	"github.com/coreos/ignition/v2/config/shared/errors"
 	"github.com/coreos/ignition/v2/config/v3_1_experimental/types"
 	"github.com/coreos/ignition/v2/internal/distro"
 	"github.com/coreos/ignition/v2/internal/exec/util"
+	"github.com/coreos/ignition/v2/internal/systemd"
 )
 
+// Preset holds the information about
+// a given systemd unit.
+type Preset struct {
+	unit            string
+	enabled         bool
+	instantiateable bool
+	instances       []string
+}
+
+// warnOnOldSystemdVersion checks the version of Systemd
+// in a given system and prints a warning if older than 240.
+func (s *stage) warnOnOldSystemdVersion() error {
+	systemdVersion, err := systemd.GetSystemdVersion()
+	if err != nil {
+		return err
+	}
+	if systemdVersion < 240 {
+		if err := s.Logger.Warning("The version of systemd (%q) is less than 240. Enabling/disabling instantiated units may not work. See https://github.com/coreos/ignition/issues/586 for more information.", systemdVersion); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
 // createUnits creates the units listed under systemd.units.
 func (s *stage) createUnits(config types.Config) error {
-	enabledOneUnit := false
+	presets := make(map[string]*Preset)
 	for _, unit := range config.Systemd.Units {
 		if err := s.writeSystemdUnit(unit, false); err != nil {
 			return err
 		}
 		if unit.Enabled != nil {
+			identifier := "disabled"
 			if *unit.Enabled {
-				if err := s.Logger.LogOp(
-					func() error { return s.EnableUnit(unit) },
-					"enabling unit %q", unit.Name,
-				); err != nil {
+				identifier = "enabled"
+			}
+			if strings.Contains(unit.Name, "@") {
+				unitName, instance, err := parseInstanceUnit(unit)
+				if err != nil {
 					return err
 				}
-			} else {
-				if err := s.Logger.LogOp(
-					func() error { return s.DisableUnit(unit) },
-					"disabling unit %q", unit.Name,
-				); err != nil {
-					return err
+				if *unit.Enabled {
+
 				}
+				unitName = fmt.Sprintf("%s-%s", unitName, identifier)
+				if _, ok := presets[unitName]; ok {
+					presets[unitName].instances = append(presets[unitName].instances, instance)
+				} else {
+					presets[unitName] = &Preset{unitName, *unit.Enabled, true, []string{instance}}
+				}
+			} else {
+				presets[unit.Name] = &Preset{unit.Name, *unit.Enabled, false, []string{}}
 			}
```

Comment:
there is an extraneous `if *unit.Enabled {}` in there. Also let's update all keys to have the identifer ("enabled" or "disabled") in it and use a more appropriate variable name like `key` instead of `unitName`. Something like this should do it:

```
diff --git a/internal/exec/stages/files/units.go b/internal/exec/stages/files/units.go
index 06f5ecc..f67cd2a 100644
--- a/internal/exec/stages/files/units.go
+++ b/internal/exec/stages/files/units.go
@@ -67,17 +67,15 @@ func (s *stage) createUnits(config types.Config) error {
                                if err != nil {
                                        return err
                                }
-                               if *unit.Enabled {
-
-                               }
-                               unitName = fmt.Sprintf("%s-%s", unitName, identifier)
-                               if _, ok := presets[unitName]; ok {
-                                       presets[unitName].instances = append(presets[unitName].instances, instance)
+                               key = fmt.Sprintf("%s-%s", unitName, identifier)
+                               if _, ok := presets[key]; ok {
+                                       presets[key].instances = append(presets[key].instances, instance)
                                } else {
-                                       presets[unitName] = &Preset{unitName, *unit.Enabled, true, []string{instance}}
+                                       presets[key] = &Preset{unitName, *unit.Enabled, true, []string{instance}}
                                }
                        } else {
-                               presets[unit.Name] = &Preset{unit.Name, *unit.Enabled, false, []string{}}
+                               key := fmt.Sprintf("%s-%s", unit.Name, identifier)
+                               presets[key] = &Preset{unit.Name, *unit.Enabled, false, []string{}}
                        }
                }
                if unit.Mask != nil && *unit.Mask {
```

one other thing is that we might not understand if we come back into this code why we added the extra identifier into the key. let's add a comment. It's basically needed specifically to support https://github.com/coreos/ignition/pull/934#discussion_r394568181

Reply:
it's also important to note that we don't need to actually update the `unit` value in the Presets types to be `echo@.service-enabled`. It should still just be `echo@.service`. The suggested patch above handles this.

Reply:
Fixed.